session_sync.go

```diff
@@ -0,0 +1,82 @@
+package remotedialer
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"time"
+)
+
+var errCloseSyncConnections = errors.New("sync from client")
+
+// encodeConnectionIDs serializes a slice of connection IDs
+func encodeConnectionIDs(ids []int64) []byte {
+	payload := make([]byte, 0, 8*len(ids))
+	for _, id := range ids {
+		payload = binary.LittleEndian.AppendUint64(payload, uint64(id))
+	}
+	return payload
+}
+
+// decodeConnectionIDs deserializes a slice of connection IDs
+func decodeConnectionIDs(payload []byte) ([]int64, error) {
+	if len(payload)%8 != 0 {
+		return nil, fmt.Errorf("incorrect data format")
+	}
+	result := make([]int64, 0, len(payload)/8)
+	for x := 0; x < len(payload); x += 8 {
+		id := binary.LittleEndian.Uint64(payload[x : x+8])
+		result = append(result, int64(id))
+	}
+	return result, nil
+}
+
+func newSyncConnectionsMessage(connectionIDs []int64) *message {
+	return &message{
+		id:          nextid(),
+		messageType: SyncConnections,
+		bytes:       encodeConnectionIDs(connectionIDs),
+	}
+}
+
+// sendSyncConnections sends a binary message of type SyncConnections, whose payload is a list of the active connection IDs for this session
+func (s *Session) sendSyncConnections() error {
+	_, err := s.writeMessage(time.Now().Add(SyncConnectionsTimeout), newSyncConnectionsMessage(s.activeConnectionIDs()))
+	return err
+}
+
+// lockedSyncConnections closes any session connection that is not present in the IDs received from the client
+// The session lock must be hold by the caller when calling this method
+func (s *Session) lockedSyncConnections(payload []byte) error {
+	clientIDs, err := decodeConnectionIDs(payload)
+	if err != nil {
+		return fmt.Errorf("decoding sync connections payload: %w", err)
+	}
+	serverIDs := s.activeConnectionIDs()
+	for _, id := range removedFromSlice(serverIDs, clientIDs) {
+		// Connection no longer active in the client, close it server-side
+		conn := s.lockedRemoveConnection(id)
+		if conn != nil {
+			// Using doTunnelClose directly instead of tunnelClose, omitting unnecessarily sending an Error message
+			conn.doTunnelClose(errCloseSyncConnections)
+		}
+	}
+	return nil
+}
+
+func removedFromSlice(a, b []int64) (res []int64) {
```

Comment:
Tom is correct. Sorry if this was not straightforward, I've added a comment to try to make it clearer.
Please let me know if I should still change the name, thanks.

Reply:
If I'm following the logic correctly, the operation being performed is a **set difference*: we want all elements present in slice A, which are not present in slice B. Since the function will fail if either input slice is not sorted and the function signature doesn't guarantee this constraint, I would lean towards something like `differenceOfSortedSets(a, b []int64)`

My thinking is that using the name of the set operation should make the intended behavior easy to search, and documenting the sorting requirement in the name will make it harder to miss this constraint at call sites. (The sorting requirement permits an efficient single-pass algorithm, which makes it worth the added complexity.)

Reply:
Thanks, that makes sense and I've renamed the function accordingly.